### PR TITLE
Add product type and github actions

### LINKED ISSUE #203: CI testing in Github actions
Set up github actions for

* Linting
* Code style checks
* Unit tests
* Functional tests

- [ ] Implement
- [ ] Write tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by pixelmatseriks**
Fixes all tests and cs
closes #203, #204, #205

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -13,13 +13,21 @@
     }
   ],
   "require": {
-    "typo3/cms-core": "^9.5|^10.4",
+    "typo3/cms-core": "^9.5.20|^10.4.6",
```
Comment (mabolek) [inline composer.json:16]: I suggest leaving v9 out of this. Leaving it in would mean we can't use PSR-14 events.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        name: "Run unit tests"
+        run: "composer ci:tests:unit"
+    strategy:
+      matrix:
+        composer-dependencies:
+          - highest
+        php-version:
+          - 7.4
+        typo3-version:
+          - ^9.5
```
Comment (mabolek) [inline .github/workflows/ci.yml:149]:
```suggestion
```

Suggesting to leave this out due to PSR-14 support. See comment in composer.json file.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+          export typo3DatabasePassword="root";
+          composer ci:tests:functional
+    strategy:
+      matrix:
+        composer-dependencies:
+          - highest
+        php-version:
+          - 7.4
+        typo3-version:
+          - ^9.5
```
Comment (mabolek) [inline .github/workflows/ci.yml:212]:
```suggestion
```

File: .php_cs.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    'strict_comparison' => true,
+    'strict_param' => true,
+    'string_line_ending' => true,
+    'standardize_increment' => true,
+    'ternary_operator_spaces' => true,
+    'trailing_comma_in_multiline_array' => true,
+    'trim_array_spaces' => true,
+    'unary_operator_spaces' => true,
+    'void_return' => true,
+    'whitespace_after_comma_in_array' => true,
```
Comment (mabolek) [inline .php_cs.php:173]: Suggestion: In pxa_siteimprove, I have checks for nesting level and cyclomatic complexity, which are good for keeping code complexity down. https://github.com/pixelant/pxa_siteimprove/blob/master/phpcs.xml.dist#L75-L76

File: Classes/Backend/FormDataProvider/ProductEditFormManipulation.php [hunk 5]
```diff
@@ -230,26 +233,27 @@ protected function simulateDataValues(Product $product, array $attributesSets, a
...
      * @param array $diffRow
      * @param array $defaultLanguageRow
      */
-    protected function setDiffData(array &$diffRow, array &$defaultLanguageRow)
+    protected function setDiffData(array &$diffRow, array &$defaultLanguageRow): void
     {
-        // TODO implementation
-        die(__METHOD__);
-
+        die('Not yet implemented: ' . __METHOD__);
```
Comment (mabolek) [inline Classes/Backend/FormDataProvider/ProductEditFormManipulation.php:255]: I suggest implementing a `NotImplementedException`

```
class NotImplementedException extends BadMethodCallException
{
}
```

File: Classes/Configuration/Flexform/Registry.php [hunk 6]
```diff
@@ -85,12 +84,13 @@ class Registry
...
 
     /**
-     * Register default actions
+     * Register default actions.
      */
     public function registerDefaultActions(): void
     {
-        if (!is_array($GLOBALS['TYPO3_CONF_VARS']['EXT']['pxa_product_manager']['switchableControllerActions']['items'])) {
-            $GLOBALS['TYPO3_CONF_VARS']['EXT']['pxa_product_manager']['switchableControllerActions']['items'] = [];
+        $scaKey = 'switchableControllerActions';
```
Comment (mabolek) [inline Classes/Configuration/Flexform/Registry.php:91]: Switchable controller actions are depecated in v10 and will be removed in v11. We should replace them.

File: Classes/Configuration/Flexform/StructureLoader.php [hunk 7]
```diff
@@ -81,18 +79,19 @@ public function loadActionDataStructure(array $dataStructure, ?array $actionConf
...
      * @return array
+     * @throws \RuntimeException
      */
     protected function updateDataStructureWithFlexform(array $dataStructure, string $flexformPath): array
     {
         $fullPath = GeneralUtility::getFileAbsFileName($flexformPath);
         if (!file_exists($fullPath)) {
             throw new \RuntimeException(
-                "Could not find flexform with path '$fullPath'(given path '$flexformPath')",
+                "Could not find flexform with path '${fullPath}'(given path '${flexformPath}')",
```
Comment (mabolek) [inline Classes/Configuration/Flexform/StructureLoader.php:94]:
```suggestion
                'Could not find flexform with path "' . $fullPath . '" (given path "' . $flexformPath . '")',
```

File: Classes/Domain/Model/Category.php [hunk 8]
```diff
@@ -29,18 +29,15 @@
...
- *
- *
- * @package pxa_product_manager
  * @license http://www.gnu.org/licenses/gpl.html GNU General Public License, version 3 or later
- *
  */
 class Category extends CategoryExtbase
 {
-    use AbleCacheProperties, CanCreateCollection;
+    use AbleCacheProperties;
```
Comment (mabolek) [inline Classes/Domain/Model/Category.php:39]: I suggest we refactor this and all other Able* namespaces to Can* now. It's better English.

```suggestion
    use CanCacheProperties;
```

File: Classes/Domain/Model/DTO/AbstractDemand.php [hunk 9]
```diff
@@ -1,29 +1,29 @@
...
 abstract class AbstractDemand implements DemandInterface
 {
     /**
-     * Array of PIDs storage
+     * Array of PIDs storage.
      *
      * @var array
      */
+    /** @codingStandardsIgnoreStart */
     protected ?array $storagePid = null;
```
Comment (mabolek) [inline Classes/Domain/Model/DTO/AbstractDemand.php:15]: Or should we simpley allow this in the CS?

File: Classes/Domain/Repository/ProductTypeRepository.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+use TYPO3\CMS\Extbase\Persistence\Repository;
+
+/**
+ * The repository for Filters.
+ */
+class ProductTypeRepository extends Repository
+{
+    use AbleFindByUidList;
```
Comment (mabolek) [inline Classes/Domain/Repository/ProductTypeRepository.php:37]: Here's another one of these Able*s that should be Can*s

```suggestion
    use CanFindByUidList;
```

File: Classes/Hook/FlexFormDataStructureHook.php [hunk 11]
```diff
@@ -12,25 +13,25 @@
...
     protected string $identifier = 'pxaproductmanager_pi1,list';
 
     /**
-     * Last selected action
+     * Last selected action.
      *
      * @var string
      */
+    /** @codingStandardsIgnoreStart */
     protected ?string $lastSelectedAction = null;
```
Comment (mabolek) [inline Classes/Hook/FlexFormDataStructureHook.php:33]: Again, I wonder if we should simply allow this.

File: Classes/LinkHandler/AbstractLinkHandler.php [hunk 12]
```diff
@@ -44,12 +44,14 @@
 use TYPO3\CMS\Recordlist\Tree\View\LinkParameterProviderInterface;
 
 /**
- * Link handler for page (and content) links
+ * Link handler for page (and content) links.
  */
+/** @codingStandardsIgnoreStart */
 abstract class AbstractLinkHandler extends Typo3LinkHandler implements LinkHandlerInterface, LinkParameterProviderInterface
```
Comment (mabolek) [inline Classes/LinkHandler/AbstractLinkHandler.php:50]: What does the CS check complain about here? It looks OK to me.

File: Classes/Service/LazyLoading/ProductsQueryDispatcher.php [hunk 13]
```diff
@@ -152,7 +151,7 @@ protected function queryBuilderToSql(QueryBuilder $queryBuilder): string
         foreach ($queryBuilder->getParameters() as $key => $value) {
             // prefix array keys with ':'
             //all non numeric values have to be quoted
-            $queryParameters[':' . $key] = is_numeric($value) ? $value : "'$value'";
+            $queryParameters[':' . $key] = is_numeric($value) ? $value : "'${value}'";
```
Comment (mabolek) [inline Classes/Service/LazyLoading/ProductsQueryDispatcher.php:154]: Looking at this part: `"'${value}'"`

I don't like variables in strings, no matter how they're put there.

Also it's being quoted for DB insertion, right? So it should be quoted using the query builder's escape function.

File: phpcs.xml [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    <rule ref="Squiz.Functions.GlobalFunction"/>
+
+    <!-- Methods -->
+    <rule ref="PSR2.Methods.MethodDeclaration.Underscore">
+        <exclude-pattern>Tests/Acceptance/</exclude-pattern>
+    </rule>
+
+    <!-- Metrics -->
+    <rule ref="Generic.Metrics.CyclomaticComplexity"/>
+    <rule ref="Generic.Metrics.NestingLevel"/>
```
Comment (mabolek) [inline phpcs.xml:76]: Oh, here they are. :-D

**Review by mabolek — CHANGES_REQUESTED** [review]
See comments.

PS! I really like (most of) the CS changes to the code. Makes it so much easier to read.

File: Classes/LinkHandler/AbstractLinkHandler.php (hunk 12 above)
Comment (pixelmatseriks) [reply Classes/LinkHandler/AbstractLinkHandler.php:50]: Line exceeds 120 characters

Comment (mabolek) [reply Classes/LinkHandler/AbstractLinkHandler.php:50]: Can't you break the line, or does CS complain about that too?

File: Classes/Service/LazyLoading/ProductsQueryDispatcher.php (hunk 13 above)
Comment (pixelmatseriks) [reply Classes/Service/LazyLoading/ProductsQueryDispatcher.php]: They are technically quoted before, looks like the query is just "extracted" as a "subQuery".
Think we need to take a look at this and change it when we look at task #212 anyway.
Just fixed the curly syntax now.

File: Classes/Configuration/Flexform/Registry.php (hunk 6 above)
Comment (pixelmatseriks) [reply Classes/Configuration/Flexform/Registry.php:91]: Maybe we can do it when we make changes to controller and actions instead?

File: Classes/Service/LazyLoading/ProductsQueryDispatcher.php (hunk 13 above)
Comment (mabolek) [reply Classes/Service/LazyLoading/ProductsQueryDispatcher.php]: OK

File: Classes/Configuration/Flexform/Registry.php (hunk 6 above)
Comment (mabolek) [reply Classes/Configuration/Flexform/Registry.php:91]: Sounds like the right thing to do. I'll add info about it to the issue.

**Review by mabolek — APPROVED** [review]

**Pull request merged**